FeedStoreChallenge/CoreData/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,111 @@
+//
+//  CoreDataFeedStore.swift
+//  Tests
+//
+//  Created by Valentin Kalchev (Zuant) on 18/09/20.
+//  Copyright © 2020 Essential Developer. All rights reserved.
+//
+
+import Foundation
+import CoreData
+
+public class CoreDataFeedStore: FeedStore {
+    private let container: NSPersistentContainer
+    private let context: NSManagedObjectContext
+    
+    public init(storeURL: URL, bundle: Bundle = .main) throws {
+        container = try NSPersistentContainer.load(withModelName: "Model", url: storeURL, in: bundle)
+        context = container.newBackgroundContext()
+    }
+    
+    public func retrieve(completion: @escaping RetrievalCompletion) {
+        let context = self.context
+        context.perform {
+            do {
+                let managedCache = try context.fetch(ManagedCache.fetchRequest() as NSFetchRequest<ManagedCache>)
+                
+                if !managedCache.isEmpty {
+                    let cache = managedCache.first!
+                    completion(.found(feed: cache.managedFeedImages.map({$0.localFeedImage}), timestamp: cache.timestamp!))
```

Comment:
Since you're interested in the first element, you can replace the `isEmpty` check with `first`. This change also allows you to remove the force unwrapping (!).

```suggestion
                if let cache = managedCache.first {
                    completion(.found(feed: cache.managedFeedImages.map({$0.localFeedImage}), timestamp: cache.timestamp!))
```

Reply:
Thanks for the feedback Caio! I'll continue with the integration tests